org.jacoco.core/src/org/jacoco/core/internal/instr/Companions.java

```diff
@@ -0,0 +1,253 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2016 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.instr;
+
+import java.lang.ref.WeakReference;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+import org.jacoco.core.internal.Java9Support;
+import org.jacoco.core.internal.data.CRC64;
+import org.jacoco.core.internal.flow.ClassProbesAdapter;
+import org.jacoco.core.runtime.RuntimeData;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+/**
+ * Instrumentation with creation of "companion" classes in "online" mode, which
+ * will hold state of instrumented classes. This allows to avoid addition of
+ * members to classes.
+ */
+public class Companions {
+
+	/**
+	 * {@link ClassLoader#defineClass(String, byte[], int, int)}
+	 */
+	private static final Method DEFINE_CLASS;
+
+	static {
+		try {
+			DEFINE_CLASS = ClassLoader.class.getDeclaredMethod("defineClass",
+					String.class, byte[].class, Integer.TYPE, Integer.TYPE);
+			DEFINE_CLASS.setAccessible(true);
+		} catch (NoSuchMethodException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static final String COMPANION_NAME = InstrSupport.DATAFIELD_NAME;
+
+	private static final String INIT_METHOD_NAME = InstrSupport.INITMETHOD_NAME;
+
+	/**
+	 * https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.11
```

Comment:
The actual value seems less important than explaining how the current value was chosen.    Perhaps something like this?

> This JVM specification requires this value to be less than 65535.  Additionally, the value must be low enough that we do not exceed the maximum method size for the initialize method in dynamically generated companion classes or the clinit method in offline generated companion classes.  This value can be arbitrarily small since multiple companion classes are generated as needed, so we choose a smaller value to avoid dynamically generating very large companion classes in case few classes are actually instrumented in a particular class loader.

Or perhaps put different constants with different comments in Companions vs OfflineInstrumentationCompanionAccessGenerator so that the comments can be a bit more specific/clearer.